File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalSort.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.planner.plan.nodes.physical.stream

import org.apache.flink.api.dag.Transformation
import org.apache.flink.streaming.api.transformations.OneInputTransformation
import org.apache.flink.table.api.{TableConfig, TableException}
import org.apache.flink.table.dataformat.BaseRow
import org.apache.flink.table.planner.calcite.FlinkTypeFactory
import org.apache.flink.table.planner.codegen.sort.ComparatorCodeGenerator
import org.apache.flink.table.planner.delegation.StreamPlanner
import org.apache.flink.table.planner.plan.nodes.exec.{ExecNode, StreamExecNode}
import org.apache.flink.table.planner.plan.utils.{RelExplainUtil, SortUtil}
import org.apache.flink.table.runtime.keyselector.NullBinaryRowKeySelector
import org.apache.flink.table.runtime.operators.sort.{ProcTimeSortOperator, RowTimeSortOperator}
import org.apache.flink.table.runtime.typeutils.BaseRowTypeInfo

import org.apache.calcite.plan.{RelOptCluster, RelTraitSet}
import org.apache.calcite.rel.RelFieldCollation.Direction
import org.apache.calcite.rel._
import org.apache.calcite.rel.core.Sort
import org.apache.calcite.rex.RexNode

import java.util

import scala.collection.JavaConversions._

/**
  * Stream physical RelNode for time-ascending-order [[Sort]] without `limit`.
  *
  * @see [[StreamExecRank]] which must be with `limit` order by.
  * @see [[StreamExecSort]] which can be used for testing now, its sort key can be any type.
  */
class StreamExecTemporalSort(
    cluster: RelOptCluster,
    traitSet: RelTraitSet,
    inputRel: RelNode,
    sortCollation: RelCollation)
  extends Sort(cluster, traitSet, inputRel, sortCollation)
  with StreamPhysicalRel
  with StreamExecNode[BaseRow] {

  override def producesUpdates: Boolean = false

  override def needsUpdatesAsRetraction(input: RelNode): Boolean = false

  override def consumesRetractions: Boolean = false

  override def producesRetractions: Boolean = false

  override def requireWatermark: Boolean = false

  override def copy(
      traitSet: RelTraitSet,
      input: RelNode,
      newCollation: RelCollation,
      offset: RexNode,
      fetch: RexNode): Sort = {
    new StreamExecTemporalSort(cluster, traitSet, input, newCollation)
  }

  override def explainTerms(pw: RelWriter): RelWriter = {
    pw.input("input", getInput())
      .item("orderBy", RelExplainUtil.collationToString(sortCollation, getRowType))
  }

  //~ ExecNode methods -----------------------------------------------------------

  /**
    * Returns an array of this node's inputs. If there are no inputs,
    * returns an empty list, not null.
    *
    * @return Array of this node's inputs
    */
  override def getInputNodes: util.List[ExecNode[StreamPlanner, _]] = {
    List(getInput.asInstanceOf[ExecNode[StreamPlanner, _]])
  }

  override def replaceInputNode(
      ordinalInParent: Int,
      newInputNode: ExecNode[StreamPlanner, _]): Unit = {
    replaceInput(ordinalInParent, newInputNode.asInstanceOf[RelNode])
  }

  override protected def translateToPlanInternal(
      planner: StreamPlanner): Transformation[BaseRow] = {
    val config = planner.getTableConfig
    // time ordering needs to be ascending
    if (SortUtil.getFirstSortDirection(sortCollation) != Direction.ASCENDING) {
      throw new TableException(
        "Sort: Primary sort order of a streaming table must be ascending on time.\n" +
          "please re-check sort statement according to the description above")
    }

    val input = getInputNodes.get(0).translateToPlan(planner)
      .asInstanceOf[Transformation[BaseRow]]

    val timeType = SortUtil.getFirstSortField(sortCollation, getRowType).getType
    timeType match {
      case _ if FlinkTypeFactory.isProctimeIndicatorType(timeType) =>
        createSortProcTime(input, config)
      case _ if FlinkTypeFactory.isRowtimeIndicatorType(timeType) =>
        createSortRowTime(input, config)
      case _ =>
        throw new TableException(
          "Sort: Internal Error\n" +
            "Normally, this happens unlikely. please contact customer support for this"
        )
    }
  }

  /**
    * Create Sort logic based on processing time
    */
  private def createSortProcTime(
      input: Transformation[BaseRow],
      tableConfig: TableConfig): Transformation[BaseRow] = {
    val inputType = FlinkTypeFactory.toLogicalRowType(getInput.getRowType)
    val fieldCollations = sortCollation.getFieldCollations
    // if the order has secondary sorting fields in addition to the proctime
    if (fieldCollations.size() > 1) {
      // strip off time collation
      val (keys, orders, nullsIsLast) = SortUtil.getKeysAndOrders(fieldCollations.tail)
      // sort code gen
      val keyTypes = keys.map(inputType.getTypeAt)
      val rowComparator = ComparatorCodeGenerator.gen(tableConfig, "ProcTimeSortComparator",
        keys, keyTypes, orders, nullsIsLast)
      val sortOperator = new ProcTimeSortOperator(BaseRowTypeInfo.of(inputType), rowComparator)
      val outputRowTypeInfo = BaseRowTypeInfo.of(FlinkTypeFactory.toLogicalRowType(getRowType))

      // as input node is singleton exchange, its parallelism is 1.
      val ret = new OneInputTransformation(
        input,
        getRelDetailedDescription,
        sortOperator,
        outputRowTypeInfo,
        input.getParallelism)

      val selector = NullBinaryRowKeySelector.INSTANCE
      ret.setStateKeySelector(selector)
      ret.setStateKeyType(selector.getProducedType)
      ret
    } else {
      // if the order is done only on proctime we only need to forward the elements
      input
    }
  }

  /**
    * Create Sort logic based on row time
    */
  private def createSortRowTime(
      input: Transformation[BaseRow],
      tableConfig: TableConfig): Transformation[BaseRow] = {
    val fieldCollations = sortCollation.getFieldCollations
    val rowTimeIdx = fieldCollations.get(0).getFieldIndex
    val inputType = FlinkTypeFactory.toLogicalRowType(getInput.getRowType)
    val rowComparator = if (fieldCollations.size() > 1) {
      // strip off time collation
      val (keys, orders, nullsIsLast) = SortUtil.getKeysAndOrders(fieldCollations.tail)
      // comparator code gen
      val keyTypes = keys.map(inputType.getTypeAt)
      ComparatorCodeGenerator.gen(tableConfig, "RowTimeSortComparator", keys, keyTypes, orders,
        nullsIsLast)
    } else {
      null
    }
    val sortOperator = new RowTimeSortOperator(
      BaseRowTypeInfo.of(inputType), rowTimeIdx, rowComparator)
    val outputRowTypeInfo = BaseRowTypeInfo.of(FlinkTypeFactory.toLogicalRowType(getRowType))

    val ret = new OneInputTransformation(
      input,
      getRelDetailedDescription,
      sortOperator,
      outputRowTypeInfo,
      input.getParallelism)

    if (inputsContainSingleton()) {
      ret.setParallelism(1)
      ret.setMaxParallelism(1)
    }

    val selector = NullBinaryRowKeySelector.INSTANCE
    ret.setStateKeySelector(selector)
    ret.setStateKeyType(selector.getProducedType)
    ret
  }
}

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecTemporalSort.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.planner.plan.nodes.exec.stream;

import org.apache.flink.api.dag.Transformation;
import org.apache.flink.streaming.api.transformations.OneInputTransformation;
import org.apache.flink.table.api.TableConfig;
import org.apache.flink.table.api.TableException;
import org.apache.flink.table.data.RowData;
import org.apache.flink.table.planner.codegen.sort.ComparatorCodeGenerator;
import org.apache.flink.table.planner.delegation.PlannerBase;
import org.apache.flink.table.planner.plan.nodes.exec.ExecEdge;
import org.apache.flink.table.planner.plan.nodes.exec.ExecNodeBase;
import org.apache.flink.table.planner.plan.nodes.exec.InputProperty;
import org.apache.flink.table.planner.plan.nodes.exec.MultipleTransformationTranslator;
import org.apache.flink.table.planner.plan.nodes.exec.spec.SortSpec;
import org.apache.flink.table.runtime.generated.GeneratedRecordComparator;
import org.apache.flink.table.runtime.keyselector.EmptyRowDataKeySelector;
import org.apache.flink.table.runtime.operators.sort.ProcTimeSortOperator;
import org.apache.flink.table.runtime.operators.sort.RowTimeSortOperator;
import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
import org.apache.flink.table.types.logical.LocalZonedTimestampType;
import org.apache.flink.table.types.logical.LogicalType;
import org.apache.flink.table.types.logical.RowType;
import org.apache.flink.table.types.logical.TimestampKind;
import org.apache.flink.table.types.logical.TimestampType;

import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Collections;
import java.util.List;

import static org.apache.flink.util.Preconditions.checkArgument;
import static org.apache.flink.util.Preconditions.checkNotNull;

/** {@link StreamExecNode} for time-ascending-order Sort without `limit`. */
public class StreamExecTemporalSort extends ExecNodeBase<RowData>
        implements StreamExecNode<RowData>, MultipleTransformationTranslator<RowData> {

    public static final String FIELD_NAME_SORT_SPEC = "orderBy";

    @JsonProperty(FIELD_NAME_SORT_SPEC)
    private final SortSpec sortSpec;

    public StreamExecTemporalSort(
            SortSpec sortSpec,
            InputProperty inputProperty,
            RowType outputType,
            String description) {
        this(
                sortSpec,
                getNewNodeId(),
                Collections.singletonList(inputProperty),
                outputType,
                description);
    }

    @JsonCreator
    public StreamExecTemporalSort(
            @JsonProperty(FIELD_NAME_SORT_SPEC) SortSpec sortSpec,
            @JsonProperty(FIELD_NAME_ID) int id,
            @JsonProperty(FIELD_NAME_INPUT_PROPERTIES) List<InputProperty> inputProperties,
            @JsonProperty(FIELD_NAME_OUTPUT_TYPE) RowType outputType,
            @JsonProperty(FIELD_NAME_DESCRIPTION) String description) {
        super(id, inputProperties, outputType, description);
        checkArgument(inputProperties.size() == 1);
        this.sortSpec = checkNotNull(sortSpec);
    }

    @SuppressWarnings("unchecked")
    @Override
    protected Transformation<RowData> translateToPlanInternal(PlannerBase planner) {
        // time ordering needs to be ascending
        if (sortSpec.getFieldSize() == 0 || !sortSpec.getFieldSpec(0).getIsAscendingOrder()) {
            throw new TableException(
                    "Sort: Primary sort order of a streaming table must be ascending on time.\n"
                            + "please re-check sort statement according to the description above");
        }

        ExecEdge inputEdge = getInputEdges().get(0);
        Transformation<RowData> inputTransform =
                (Transformation<RowData>) inputEdge.translateToPlan(planner);

        RowType inputType = (RowType) inputEdge.getOutputType();
        LogicalType timeType = inputType.getTypeAt(sortSpec.getFieldSpec(0).getFieldIndex());
        TableConfig config = planner.getTableConfig();
        if (timeType instanceof TimestampType) {
            TimestampType keyType = (TimestampType) timeType;
            if (keyType.getKind() == TimestampKind.ROWTIME) {
                return createSortRowTime(inputType, inputTransform, config);
            }
        }
        if (timeType instanceof LocalZonedTimestampType) {
            LocalZonedTimestampType keyType = (LocalZonedTimestampType) timeType;
            if (keyType.getKind() == TimestampKind.PROCTIME) {
                return createSortProcTime(inputType, inputTransform, config);
            }
        }

        throw new TableException(
                String.format(
                        "Sort: Internal Error\n"
                                + "First field in temporal sort is not a time attribute, %s is given.",
                        timeType));
    }

    /** Create Sort logic based on processing time. */
    private Transformation<RowData> createSortProcTime(
            RowType inputType, Transformation<RowData> inputTransform, TableConfig tableConfig) {
        // if the order has secondary sorting fields in addition to the proctime
        if (sortSpec.getFieldSize() > 1) {
            // skip the first field which is the proctime field and would be ordered by timer.
            SortSpec specExcludeTime = sortSpec.createSubSortSpec(1);

            GeneratedRecordComparator rowComparator =
                    ComparatorCodeGenerator.gen(
                            tableConfig, "ProcTimeSortComparator", inputType, specExcludeTime);
            ProcTimeSortOperator sortOperator =
                    new ProcTimeSortOperator(InternalTypeInfo.of(inputType), rowComparator);

            OneInputTransformation<RowData, RowData> transform =
                    new OneInputTransformation<>(
                            inputTransform,
                            getDescription(),
                            sortOperator,
                            InternalTypeInfo.of(inputType),
                            inputTransform.getParallelism());

            // as input node is singleton exchange, its parallelism is 1.
            if (inputsContainSingleton()) {
                transform.setParallelism(1);
                transform.setMaxParallelism(1);
            }

            EmptyRowDataKeySelector selector = EmptyRowDataKeySelector.INSTANCE;
            transform.setStateKeySelector(selector);
            transform.setStateKeyType(selector.getProducedType());
            return transform;
        } else {
            // if the order is done only on proctime we only need to forward the elements
            return inputTransform;
        }
    }

    /** Create Sort logic based on row time. */
    private Transformation<RowData> createSortRowTime(
            RowType inputType, Transformation<RowData> inputTransform, TableConfig tableConfig) {
        GeneratedRecordComparator rowComparator = null;
        if (sortSpec.getFieldSize() > 1) {
            // skip the first field which is the rowtime field and would be ordered by timer.
            SortSpec specExcludeTime = sortSpec.createSubSortSpec(1);
            rowComparator =
                    ComparatorCodeGenerator.gen(
                            tableConfig, "RowTimeSortComparator", inputType, specExcludeTime);
        }
        RowTimeSortOperator sortOperator =
                new RowTimeSortOperator(
                        InternalTypeInfo.of(inputType),
                        sortSpec.getFieldSpec(0).getFieldIndex(),
                        rowComparator);

        OneInputTransformation<RowData, RowData> transform =
                new OneInputTransformation<>(
                        inputTransform,
                        getDescription(),
                        sortOperator,
                        InternalTypeInfo.of(inputType),
                        inputTransform.getParallelism());

        if (inputsContainSingleton()) {
            transform.setParallelism(1);
            transform.setMaxParallelism(1);
        }

        EmptyRowDataKeySelector selector = EmptyRowDataKeySelector.INSTANCE;
        transform.setStateKeySelector(selector);
        transform.setStateKeyType(selector.getProducedType());
        return transform;
    }
}
